### PR TITLE
Fix support for multiple additionalNetworkCIDR blocks

### LINKED ISSUE #8228: Wrong terraform code generated for additional network CIDRs
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
v1.15.0

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
v1.15.6

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
I updated my cluster.yml with following values
```
additionalNetworkCIDRs:
    - 172.63.0.0/16
    - 172.64.0.0/16
    - 172.65.0.0/16
```
and then ran
kops replace cluster -f <cluster.name>/cluster.yaml --state=s3://store
kops update cluster <cluster.name> --state s3://store --out=<cluster.name>/terraform/kops

**5. What happened after the commands executed?**
There were some `resource "aws_vpc_ipv4_cidr_block_association"` added for all the three additional CIDR blocks
```
resource "aws_vpc_ipv4_cidr_block_association" "172-63-0-0--16" {
  vpc_id     = "${aws_vpc.cluster-name.id}"
  cidr_block = "172.65.0.0/16"
}
resource "aws_vpc_ipv4_cidr_block_association" "172-64-0-0--16" {
  vpc_id     = "${aws_vpc.cluster-name.id}"
  cidr_block = "172.65.0.0/16"
}
resource "aws_vpc_ipv4_cidr_block_association" "172-65-0-0--16" {
  vpc_id     = "${aws_vpc.cluster-name.id}"
  cidr_block = "172.65.0.0/16"
}
```

**6. What did you expect to happen?**
All the CIDR block association terraform code generated by kops has same `cidr_block`

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
```
---
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 
  name: cluster.name
spec:
  api:
    loadBalancer:
      type: Internal
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    Stage: prod
    cluster_name: cluster.name
    environment: spike
  cloudProvider: aws
  configBase: s3://store
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: master-ap-south-1a
      name: a
    - instanceGroup: master-ap-south-1b
      name: b
    - instanceGroup: master-ap-south-1c
      name: c
    memoryRequest: 100Mi
    name: main
    version: 3.3.10
    provider: Legacy
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: master-ap-south-1a
      name: a
    - instanceGroup: master-ap-south-1b
      name: b
    - instanceGroup: master-ap-south-1c
      name: c
    memoryRequest: 100Mi
    name: events
    version: 3.3.10
    provider: Legacy
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: true
    kubeletCgroups: "/podruntime.slice"
    runtimeCgroups: "/podruntime.slice"
    clusterDNS: 169.254.20.11,PRIVATE_EC2_IPV4
    evictionHard: memory.available<750Mi,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5%
    kubeReserved:
      cpu: 300m
      memory: 1.0Gi
    kubeReservedCgroup: "/podruntime.slice"
    enforceNodeAllocatable: pods,kube-reserved
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.15.6
  masterPublicName: api.cluster.name
  networkCIDR: 172.62.0.0/16
  additionalNetworkCIDRs:
    - 172.63.0.0/16
    - 172.64.0.0/16
    - 172.65.0.0/16
  networking:
    amazonvpc: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.62.0.0/17
    name: ap-south-1a
    type: Private
    zone: ap-south-1a
  - cidr: 172.62.128.0/17
    name: ap-south-1b
    type: Private
    zone: ap-south-1b
  - cidr: 172.63.0.0/17
    name: ap-south-1c
    type: Private
    zone: ap-south-1c
  - cidr: 172.63.128.0/19
    name: utility-ap-south-1a
    type: Utility
    zone: ap-south-1a
  - cidr: 172.63.160.0/19
    name: utility-ap-south-1b
    type: Utility
    zone: ap-south-1b
  - cidr: 172.63.192.0/19
    name: utility-ap-south-1c
    type: Utility
    zone: ap-south-1c
  - cidr: 172.63.224.0/21
    name: secure-ap-south-1a
    type: Private
    zone: ap-south-1a
  - cidr: 172.65.232.0/21
    name: secure-ap-south-1b
    type: Private
    zone: ap-south-1b
  - cidr: 172.65.240.0/21
    name: secure-ap-south-1c
    type: Private
    zone: ap-south-1c
  topology:
    bastion:
      bastionPublicName: bastion.cluster.name
    dns:
      type: Public
    masters: private
    nodes: private
  fileAssets:
  - name: podruntime-slice
    path: "/usr/lib/systemd/system/podruntime.slice"
    content: |
      [Unit]
      Description=Limited resources slice for Kubernetes services
      Documentation=man:systemd.special(7)
      DefaultDependencies=no
      Before=slices.target
      Requires=-.slice
      After=-.slice
  - name: resource-tagging
    path: "/tmp/kops/resource-tagging.sh"
    content: |
      DEVICE=$(curl -s  http://instance-data/latest/meta-data/block-device-mapping/root/)
      INSTANCE_ID=$(curl -s  http://instance-data/latest/meta-data/instance-id)
      REGION=$(curl -s http://instance-data/latest/dynamic/instance-identity/document | jq .region -r)
      VOLUME_ID=$(aws ec2 describe-instances --instance-ids=$INSTANCE_ID  --region=$REGION  | jq -r '.Reservations[0].Instances[0].BlockDeviceMappings[] | select (.DeviceName =="'${DEVICE}'" )| .Ebs.VolumeId')
      TAGS=$(aws ec2 describe-tags --filters Name=resource-id,Values=$INSTANCE_ID  --region $REGION --output=text --query='Tags[*].[Key, Value]' | grep -v '^aws:' | awk '{printf " Key="$1",Value="$2 }')
      aws ec2 create-tags --resource=$VOLUME_ID --tags $TAGS --region=$REGION
  hooks:
  - before:
    - kubelet.service
    manifest: |
      [Unit]
      Description=Set PRIVATE_EC2_IPV4 cluster-dns
      [Service]
      ExecStart=/bin/bash -c "sed -i 's/PRIVATE_EC2_IPV4/'$(/usr/bin/curl --silent --fail http://169.254.169.254/latest/meta-data/local-ipv4)'/g' /etc/sysconfig/kubelet; sudo sed -i '1s/"$/ --node-ip='$(/usr/bin/curl --silent --fail    http://169.254.169.254/latest/meta-data/local-ipv4)'"/g' /etc/sysconfig/kubelet; /bin/systemctl daemon-reload; /bin/systemctl restart kubelet"
      RemainAfterExit=yes
    name: private-ipv4
  - after:
    - kubelet.service
    manifest: |
      [Unit]
      Description=Tag Root Volume
      [Service]
      ExecStart=/bin/bash -x /tmp/kops/resource-tagging.sh
      RemainAfterExit=yes
    name: resources-tagging
  - before:
    - kubelet.service
    - docker.service
    manifest: |
      [Unit]
      Description=Tune nofile ulimit
      [Service]
      ExecStart=/bin/bash -c "echo -e 'fs.file-max = 200000000\nkernel.pid_max = 3500000' >> /etc/sysctl.conf && sudo sysctl -p && echo -e '* soft nofile 200000\n* hard nofile 200000' >> /etc/security/limits.conf && systemctl set-property docker.service TasksMax=infinity"
      RemainAfterExit=yes
    name: nofile-ulimit
  authentication:
    aws: {}
  docker:
    defaultUlimit:
    - nofile=1000000:1000000
    version: 18.06.2
  kubeDNS:
    provider: CoreDNS
  cloudConfig:
    disableSecurityGroupIngress: true
  kubeAPIServer:
    serviceClusterIPRange: 100.64.0.0/13
    admissionControl:
    - PodPreset
    runtimeConfig:
      settings.k8s.io/v1alpha1: 'true'
---
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 
  labels:
    kops.k8s.io/cluster: cluster.name
  name: master-ap-south-1a
spec:
  image: '099720109477/ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-20180405'
  machineType: t3.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-ap-south-1a
  role: Master
  subnets:
  - ap-south-1a
---
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 
  labels:
    kops.k8s.io/cluster: cluster.name
  name: master-ap-south-1b
spec:
  image: '099720109477/ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-20180405'
  machineType: t3.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-ap-south-1b
  role: Master
  subnets:
  - ap-south-1b
---
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 
  labels:
    kops.k8s.io/cluster: cluster.name
  name: master-ap-south-1c
spec:
  image: '099720109477/ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-20180405'
  machineType: t3.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-ap-south-1c
  role: Master
  subnets:
  - ap-south-1c
---
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 
  labels:
    kops.k8s.io/cluster: cluster.name
  name: nodes
spec:
  image: '099720109477/ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-20180405'
  machineType: t2.medium
  maxSize: 3
  minSize: 3
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - ap-south-1a
  - ap-south-1b
  - ap-south-1c
---
apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 
  labels:
    kops.k8s.io/cluster: cluster.name
  name: bastions
spec:
  image: '099720109477/ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-20180405'
  machineType: t2.micro
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: bastions
  role: Bastion
  subnets:
  - utility-ap-south-1a
  - utility-ap-south-1b
  - utility-ap-south-1c
```
  You may want to remove your cluster name and other sensitive information.**

### PR DESCRIPTION
**Pull request by rifelpet**
Fixes #8228

### CODE REVIEW

**Comment by rifelpet** [comment]
This is ready for review. In order to disassociate associations that are no longer in the `additionalNetworkCIDRs` list I had to mimic the behavior of the SecurityGroup rule removal.

We now pass a list of all CIDR blocks that _should_ be associated with the VPC and disassociate any that we find outside that list.

manual test sequence I performed:

1. 
    1. Add new `additionalNetworkCIDRs` list item.
    2. `update` reports new VPCCIDRBlock resource.
    3. `update --yes` adds the CIDR to the VPC.
    4. Fresh `update` reports no changes.
2. 
    1. Remove the CIDR from `additionalNetworkCIDRs`.
    2. `update` reports deletion.
    3. `update --yes` disassociates the CIDR from the VPC.
    4. The association still exists but has a state of `disassociated`.
    5. Fresh `update` reports no changes.
3.
    1. Readd the CIDR back to `additionalNetworkCIDRs`.
    2. `update` reports new VPCCIDRBlock resource.
    3. `update --yes` creates a new association alongside the disassociated association.
    4. Fresh `update` reports no changes


Note: the association list includes the primary CIDR, so we need to ignore that when building the list of associations to disassociate.

example deletion output during `update`:
```
Will delete items:
  VPCCIDRBlock         vpc-09acd439a582f8e95: cidr=172.22.0.0/16
```

example list from `describeVPCs` output, showing multiple associations with the same cidr block but different states:

```json
            "CidrBlockAssociationSet": [
                {
                    "AssociationId": "vpc-cidr-assoc-0f599dd6995dfcd92",
                    "CidrBlock": "172.20.0.0/16",
                    "CidrBlockState": {
                        "State": "associated"
                    }
                },
                {
                    "AssociationId": "vpc-cidr-assoc-06297d26c880c276c",
                    "CidrBlock": "172.21.0.0/16",
                    "CidrBlockState": {
                        "State": "associated"
                    }
                },
                {
                    "AssociationId": "vpc-cidr-assoc-059d42dcf8cdd2dc8",
                    "CidrBlock": "172.22.0.0/16",
                    "CidrBlockState": {
                        "State": "disassociated"
                    }
                },
                {
                    "AssociationId": "vpc-cidr-assoc-0002d4de834a60151",
                    "CidrBlock": "172.22.0.0/16",
                    "CidrBlockState": {
                        "State": "associated"
                    }
                }
            ],
```

File: upup/pkg/fi/cloudup/awstasks/vpc.go [hunk 1]
```diff
@@ -44,9 +44,14 @@ type VPC struct {
 	Shared *bool
 
 	Tags map[string]string
+
+	// RemoveExtraCIDRBlocks contains a list of cidr blocks that should be
+	// associated with the VPC; any other CIDR blocks should be disassociated.
+	RemoveExtraCIDRBlocks []string
```
Comment (olemarkus) [inline upup/pkg/fi/cloudup/awstasks/vpc.go:50]: To me, the name of the variable suggests it contains which CIDRs to delete. Maybe `AssociateExtraCIDRBlocks` or `KeepCIDRBlocks` are better names?

Comment (rifelpet) [reply upup/pkg/fi/cloudup/awstasks/vpc.go]: Agreed the variable name is not ideal, I just updated to `AssociateExtraCIDRBlocks`. For reference I was following how SecurityGroup handles it:
https://github.com/kubernetes/kops/blob/b9921511a46a4b270073e18a676520cbd819f243/upup/pkg/fi/cloudup/awstasks/securitygroup.go#L42

but we should probably update that variable as well.

**Review by hakman — APPROVED** [review]

**Comment by codablock** [comment]
@rifelpet Looks like this PR causes shared VPCs to not work anymore in combination with additionalNetworkCIDR. It tries to disassociate CIDRs from the shared VPC, while these are externally managed (by our own Terraform). I assume the new method `FindDeletions` in vpc.go should check for shared VPCs first. I'm able to help myself for now by simple returning `nil, nil` from that method for now. Unfortunately I don't feel confident enough to create a proper PR for this one, as it would require good tests which seem to be quite complicated.

**Comment by rifelpet** [comment]
@codablock good catch! I can submit a fix shortly. we don't have very solid test coverage of shared networking setups (especially making this type of assertion) but I'll see if we can get something in place

**Pull request merged**